### PR TITLE
Generate Neon scalars

### PR DESCRIPTION
**Pull request by murzinv**
Hi All,

I've been adding capability to generate Neon load/store pair instruction and Neon load/store scalar (including ordered variants `LDAPUR` and `STLUR`). Generator now accepts new psudo relaxations:
- `NePa` for Neon `LDP`/`STP`
- `NePaN` for Neon `LDNP`/`STNP`
- `NeP` for Neon `LDUR`/`STUR` instructions
- `NeQ` and `NeL` for Neon `LDAPUR` and `STLUR` correspondingly

Additionally, identified some places for improvements and added support for Neon `LDAP1` and `STL1` in `herd` and `litmus`. 

Please, note that new Neon ordered instructions might not be supported by compilers yet.

### CODE REVIEW

**Comment by relokin** [comment]
Some minor comments and I think it would be good to add single instruction tests to `herd/test/instructions` for the newly added instructions. Otherwise, LGTM!

**Review by relokin — COMMENTED** [review]
Sorry, hope you can see them now...

**Comment by relokin** [comment]
> > Some minor comments and I think it would be good to add single instruction tests to `herd/test/instructions` for the newly added instructions. Otherwise, LGTM!
> 
> Thanks @relokin for having a look! Could you share your comments, please? I'm happy to add tests, the only reason I have not done it yet because tests in `herd/test/instructions/AArcch64.neon` can be consumed by `litmus7` and support for new instructions is on-going [1], so there is no way to build tests which have new instructions. Now, I'm wondering if there is a way inform test framework about current limitation?
> 
> [1] https://sourceware.org/pipermail/binutils/2024-January/131805.html

There are more tests in `herd/test/instructions/` that can't be handled by litmus7 (for example all MTE), not ideal but it's happened before. In this case, the divergence is due to compiler limitations (not litmus7), so I think having these tests makes sense. In any case, the regressions do not test litmus7 yet so one would have to manually run.

**Comment by maranget** [comment]
> There are more tests in `herd/test/instructions/` that can't be handled by litmus7 (for example all MTE), not ideal but it's happened before. In this case, the divergence is due to compiler limitations (not litmus7), so I think having these tests makes sense. In any case, the regressions do not test litmus7 yet so one would have to manually run.

Maybe you can create a new directory `herd/tests/instructions/AArch64.SVE` for the new tests that rely on the new SVE instructions, instead of adding the new tests to the ones in `herd/tests/instructions/AArch64.neon` ?

**Comment by hugookeeffe** [comment]
I've gone through the code for gen, and it all looks good as far as I can tell. LGTM! thank you for the patch @murzinv

File: gen/AArch64Arch_gen.ml [hunk 1]
```diff
@@ -60,49 +60,67 @@ module Mixed =
...
-  let pp n = Printf.sprintf "Ne%i%s" (nregs n) (pp_opt n)
+  let pp n =
+    match n with
+    | Ne1 | Ne2 | Ne3 | Ne4 | Ne2i | Ne3i | Ne4i ->
+       Printf.sprintf "Ne%i%s" (nregs n) (pp_opt n)
+    | NePa -> Printf.sprintf "NePa"
+    | NePaN -> Printf.sprintf "NePaN"
+    | NeP -> Printf.sprintf "NeP"
+    | NeAcqPc -> Printf.sprintf "NeQ"
+    | NeRel -> Printf.sprintf "NeL"
```
Comment (relokin) [inline gen/AArch64Arch_gen.ml:91]: The use of sprintf is not necessary is it?
```suggestion
    | NePa -> "NePa"
    | NePaN -> "NePaN"
    | NeP -> "NeP"
    | NeAcqPc -> "NeQ"
    | NeRel -> "NeL"
```

**Review by relokin — APPROVED** [review]
One minor comment otherwise the change LGTM! Let's give it some time in case anyone else wants to review. If you address the glitch and rebase we can merge by the end of this week.

**Comment by relokin** [comment]
We're planning to merge this later today, please let us know if there are any objections or you need more time to review this.

**Pull request merged**